### PR TITLE
Enhancement on Substrate Node Template

### PR DESCRIPTION
**Pull request by jimmychu0807**
- Updated Substrate Node Template version to `3.0.0`
- `spec_version` in runtime is set to 100
- Added a documentation on how to release Substrate Node Template
- Misc minor enhancement
- Copying documentations that are currently in [Substrate Node Template]() back to here (upstream)
- The `node-template-release.sh` script is directly using the path passed in instead of forcing to use a relative path.

### CODE REVIEW

File: bin/node-template/README.md [hunk 1]
```diff
@@ -1,97 +1,71 @@
...
-cargo run -- \
-  --base-path /tmp/bob \
-  --bootnodes /ip4/127.0.0.1/tcp/30333/p2p/12D3KooWEyoppNCUx8Yx66oV9fJnriXwCcXwDDUA2kj6vnc6iDEp \
-  --chain=local \
-  --bob \
-  --port 30334 \
-  --ws-port 9945 \
-  --telemetry-url 'ws://telemetry.polkadot.io:1024 0' \
-  --validator
+RUST_LOG=debug RUST_BACKTRACE=1 ./target/release/node-template -lruntime=debug --dev
```
Comment (kianenigma) [inline bin/node-template/README.md:62]: FYI we also support `-ldebug`. Basically anything that you put in front of RUST_LOG can go with -l

File: bin/node-template/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+version: "3.2"
```
Comment (kianenigma) [inline bin/node-template/docker-compose.yml:1]: I don't have anything against it but why a docker compose for a single service? 

Now that we have a compose file, we could bundle this with a UI image as well.

File: bin/node-template/runtime/src/lib.rs [hunk 3]
```diff
@@ -96,7 +96,8 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
 	spec_name: create_runtime_str!("node-template"),
 	impl_name: create_runtime_str!("node-template"),
 	authoring_version: 1,
-	spec_version: 1,
+	// TODO: adding reason on setting spec_version to 100.
```
Comment (kianenigma) [inline bin/node-template/runtime/src/lib.rs:99]: yes would be very good.

File: bin/node-template/scripts/init.sh [hunk 4]
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env bash
-
+# This script meant to be run on Unix/Linux based systems
```
Comment (kianenigma) [inline bin/node-template/scripts/init.sh:2]:
```suggestion
# This script is meant to be run on Unix/Linux based systems
```

**Review by kianenigma — APPROVED** [review]
Looks overall good to me.

File: bin/node-template/docker-compose.yml (hunk 2 above)
Comment (jimmychu0807) [reply bin/node-template/docker-compose.yml:1]: Indeed. Let me deal with this separately in this issue:
https://github.com/substrate-developer-hub/substrate-node-template/issues/163

File: bin/node-template/runtime/src/lib.rs [hunk 5]
```diff
@@ -92,11 +92,18 @@ pub mod opaque {
...
 	spec_name: create_runtime_str!("node-template"),
 	impl_name: create_runtime_str!("node-template"),
 	authoring_version: 1,
-	spec_version: 1,
+	// The version of the runtime specification. A full node will not attempt to use its native
+	//   runtime in substitute for the on-chain Wasm runtime unless all of `spec_name`,
+	//   `spec_version`, and `authoring_version` are the same between Wasm and native.
+	// This value is set to 100 to notify Polkadot-JS App (https://polkadot.js.org/apps) to use
+	//   the compatible custom types on the client side.
+	spec_version: 100,
```
Comment (jimmychu0807) [inline bin/node-template/runtime/src/lib.rs:106]: adding explanation per https://github.com/paritytech/substrate/pull/8473/#discussion_r602849475

File: bin/node-template/runtime/src/lib.rs (hunk 3 above)
Comment (jimmychu0807) [reply bin/node-template/runtime/src/lib.rs]: Added below.

**Comment by jimmychu0807** [comment]
thanks @kianenigma for the review. Addressed all of your concerns, and about docker script I will deal with it in a separate issue.

File: docs/node-template-release.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+	cd substrate-node-template
+	# rsync with force copying. Note the slash at the destination directory is important
+	rsync -avh * <destination node-template directory>/
+	# For dry-running add `-n` argument
+	# rsync -avhn * <destination node-template directory>/
+	```
+
+	The above command only copy existing files from the source to destination, but not deleting
+	files/directories that are removed from the source. So manually check and remove them in the
+	destination.
```
Comment (apopiak) [inline docs/node-template-release.md:26]:
```suggestion
	The above command only copies existing files from the source to the destination, but does not delete
	files/directories that are removed from the source. So you need to manually check and remove them in the
	destination.
```

File: docs/node-template-release.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
+# Node Template Release Process
+
+1. It has to be in a github checkout directory with your current work committed into
+`https://github.com/paritytech/substrate/`, because this is where the build script will check out
+from. Assume you are in root directory of Substrate. Run:
```
Comment (apopiak) [inline docs/node-template-release.md:5]: this is confusing to me, can you rephrase it?

File: docs/node-template-release.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+		[dev-dependencies.sp-core]
+		default-features = false
+		git = 'https://github.com/paritytech/substrate.git'
+		rev = 'c1fe59d060600a10eebb4ace277af1fee20bad17'
+		version = '3.0.0'
+		```
+
+		We will update each of them to the shortened form and linked to the
+		[crate registry](https://crates.io/). After confirming the versioned package is published in
+		crate, the above will become:
```
Comment (apopiak) [inline docs/node-template-release.md:43]:
```suggestion
		We will update each of them to the shortened form and linked to the Rust
		[crate registry](https://crates.io/). After confirming the versioned package is published in
		the crate, the above will become:
```

File: docs/node-template-release.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+		```toml
+		[dev-dependencies]
+		sp-core = { version = '3.0.0', default-features = false }
+		```
+
+	P.S: This step can be automated if we update `node-template-release` package in
+	`.maintain/node-template-release`.
```
Comment (apopiak) [inline docs/node-template-release.md:51]: this sounds like a todo?

File: docs/node-template-release.md [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+4. Once the three `Cargo.toml`s are updated, compile and confirm Node Template builds. Then
+commit the changes to a new branch in [Substrate Node Template](https://github.com/substrate-developer-hub/substrate-node-template), and make a PR.
```
Comment (apopiak) [inline docs/node-template-release.md:54]:
```suggestion
4. Once the three `Cargo.toml`s are updated, compile and confirm that the Node Template builds. Then
commit the changes to a new branch in [Substrate Node Template](https://github.com/substrate-developer-hub/substrate-node-template), and make a PR.
```

File: docs/node-template-release.md [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+	> Notice there are chances the code in Substrate Node Template works with the linked Substrate git
+	commit, but not with published packages due to the latest unpublished (yet) features. In this case,
+	resume that section of Node Template to its previous version to ensure Node Template builds.
```
Comment (apopiak) [inline docs/node-template-release.md:58]:
```suggestion
	> Note that there is a chance the code in Substrate Node Template works with the linked Substrate git
	commit but not with published packages due to the latest (as yet) unpublished features. In this case,
	rollback that section of the Node Template to its previous version to ensure the Node Template builds.
```

**Review by apopiak — COMMENTED** [review]
awesome that you're documenting the release process!

File: docs/node-template-release.md (hunk 9 above)
Comment (jimmychu0807) [reply docs/node-template-release.md:51]: It is. But given there are 626 issues in Substrate repo already, I don't want to add another one that is kind of low priority.

File: docs/node-template-release.md (hunk 7 above)
Comment (jimmychu0807) [reply docs/node-template-release.md]: @apopiak this paragraph is updated.

**Comment by jimmychu0807** [comment]
@apopiak thanks for the review 🙏

File: docs/node-template-release.md (hunk 9 above)
Comment (apopiak) [reply docs/node-template-release.md:53]: I don't think it makes sense not to create issues and have things in code, but :shrug:

**Review by apopiak — APPROVED** [review]

**Review by s3krit — APPROVED** [review]
Awesome to see we're documenting this more. I'll need to look into this a lot more to think about how we can better improve the automation for releasing node-template in the future, but in the meantime this is a really good change. Glad to see it's being documented :)

File: bin/node-template/.editorconfig [hunk 12]
```diff
@@ -0,0 +1,16 @@
+root = true
+
+[*]
+indent_style=space
+indent_size=2
+tab_width=2
```
Comment (shawntabrizi) [inline bin/node-template/.editorconfig:6]: did we want to commit this file?
